src/lazy/streaming_raw_reader.rs

```diff
@@ -83,11 +108,13 @@ impl<Encoding: Decoder, Input: IonInput> StreamingRawReader<Encoding, Input> {
 
             let bytes_read = end_position - starting_position;
             let input = unsafe { &mut *self.input.get() };
-            // If we've exhausted the buffer...
-            if bytes_read >= available_bytes.len() || matches!(result, Err(IonError::Incomplete(_)))
-            {
-                // ...try to pull more data from the data source. If we get more data, try again.
```

Comment:
🗺️ This was a subtle bug enabled by my use of `unsafe`. If `result` was `Ok(_)` but _also_ consumed all of the data in `available_bytes`, then this code path would have `input` shift its bytes to the beginning and pull in more data. Because `result` contained an `Ok(value)` referring to a slice of that buffer, its data would get clobbered.